File: CPP VERSION/src/Level_selection.cpp
#include "Loops.hpp"
#include <sys/stat.h>
#include <unistd.h>

inline bool exists (const std::string& name) {
  struct stat buffer;
  return (stat (name.c_str(), &buffer) == 0);
}

void selection_menu(sf::RenderWindow &app) {
    Background bg("background_menu.png");
    Cursor mouse;
    sf::Vector2i mousePos;
    sf::Event event;

    std::vector<Button> buttons;
    std::vector<bool> unlocked = get_unlocked();

    Message title(512, 50, 36, "Select a level", "impact.ttf", sf::Color(200,0,0));
    Button back(512, 590, 36, "Back", "impact.ttf", sf::Color(200,0,0));
    Message help_msg(512, 620, 18 ,"You can only select levels you've unlocked. To unlock a level, beat all the previous ones.", "impact.ttf", sf::Color(69,52,16));
    // load the buttons
    int n,i,j;
    for (n = 1 ; n<=NB_LEVEL ; n++) {
        if (i>500) {
            i=0;
            j+=50;
        }
        buttons.emplace_back(100+j,100+i, 20, "#"+std::to_string(n), "impact.ttf" ,sf::Color(200,0,0));
        i+=30;
    }
    int n_customs = n-1;
    int m = 1; i = 0; j = 0;
    while (exists("./static/custom_levels/custom_level"+std::to_string(m)+".txt")) {
        if (i>500) {
            i = 0;
            j+=50;
        }
        buttons.emplace_back(900+j, 100+i, 20, "#custom"+std::to_string(m), "impact.ttf" ,sf::Color(200,0,0));
        i+=30;
        m++;
    }

    while (app.isOpen()) {

        mousePos = sf::Mouse::getPosition(app);

        while(app.pollEvent(event)) {
            // check the type of the event...
            switch (event.type) {
               // window closed
               case sf::Event::Closed:
                   app.close();
                   return;

               // key pressed
               case sf::Event::KeyPressed: {
                   if(sf::Keyboard::Escape)
                       return;
                   break;
               }

              case sf::Event::MouseButtonPressed :
                  if (event.mouseButton.button == sf::Mouse::Left) {
                       if (back.is_lit())
                           return;
                       else {
                           for (int i = 0 ; i < int(buttons.size()); i++) {
                               if (buttons[i].is_lit()) {
                                   menu_music.stop();
                                   if (i>=n_customs) {
                                       bool x = level_loop(i-n_customs+1, true, true, true, app);
                                   } else if (unlocked[i]) {
                                       if (level_loop(i+1, false, true, true, app))
                                            unlocked = get_unlocked();
                                   }
                                   main_music.stop();
                                   menu_music.play();
                               }
                           }
                       }
                  }
                  break;

               // we don't process other types of events
               default:
                   break;
            }
        }

        bg.update(app);
        title.update(app);
        help_msg.update(app);
        back.update(mousePos.x, mousePos.y, app);
        for (int i = 0 ; i < int(buttons.size()); i++) {
            buttons[i].update(mousePos.x, mousePos.y, app);
        }

        mouse.update(mousePos.x, mousePos.y, app);
        app.display();
    }
}

File: PYTHON VERSION/src/Level_Selection.py
# Tank Game:
# A game with tanks and stuff.
#
# by Erkalys & Florimond Manca
#
# Level selection menu


## Imports

import pygame
from pygame.locals import *
from pygame.font import *
import math
import sys
import os
import Tank
import Utils
import Bullet_Cursor
from paths import *
from AI_Player import *
from Level_Loop import main
from Importation import get_level, get_unlocked, overwrite_unlocked


def button_update(buttons, bliton = pygame.display.get_surface()):
    for b in buttons:
        b.update(bliton)

def level_selection_menu():
    """level_selection_menu(): launches the menu where the player can select
    levels, either normal or custom ones."""
    # some initialization
    pygame.font.init()  # pygame module for managing fonts
    size = Utils.get_size()
    screen = pygame.display.set_mode(size)
    background = Utils.Background(IMAGE_PATH, nlevel=-1)
    screen.blit(background.image, (0, 0))

    # replace the mouse with a cursor
    pygame.mouse.set_visible(False)
    curseur = Bullet_Cursor.Cursor(IMAGE_PATH)

    clickSound = pygame.mixer.Sound(join(MUSIC_PATH, "click_sound.wav"))

    clock = pygame.time.Clock()
    running = 1
    buttons = []
    unlocked = get_unlocked() # unlocked levels
    font_big = pygame.font.SysFont("impact", 36)
    font_medium = pygame.font.SysFont("impact", 18)
    font_buttons = pygame.font.SysFont("impact", 20)

    title = font_big.render("Select a level", 1, (30, 30, 30))
    titlepos = title.get_rect(centerx=512, centery=50)

    back = Utils.Button("Back" , font_big, 512, 590, (200, 0, 0))
    help_msg = font_medium.render("You can only select levels you've unlocked. To unlock a level, beat all the previous ones.", 1, (69, 52, 16))
    msgpos = help_msg.get_rect(centerx=512, centery=620)

    # load the buttons
    n, i, j = 1, 0, 0
    # first the buttons for the default levels
    while os.path.exists(join(LEVEL_PATH, "level"+str(n)+".txt")):
        if i > 500:
            i = 0
            j += 50
        buttons.append(Utils.Button("#{}".format(n), font_buttons,
            100 + j, 100 + i, (200, 0, 0)))
        i += 30
        n += 1
    n_customs = n - 1 # the identifier of the first custom level

    m, i, j = 1, 0, 0

    while os.path.exists(join(CLEVEL_PATH, "custom_level{}.txt".format(m))): #on charge les boutons des niveaux customs
        if i > 500:
            i = 0
            j += 50
        buttons.append(Utils.Button("#custom{}".format(m), font_buttons,
            900 + j, 100 + i, (200, 0, 0)))
        i += 30
        m += 1

    # On lance la boucle d'affichage
    while running:
        clock.tick(60)  # maximum fps is 60
        # blit the basic things
        screen.blit(background.image , (0, 0))
        screen.blit(title, titlepos)
        screen.blit(help_msg, msgpos)
        # update the entities
        back.update(screen)
        button_update(buttons,screen)
        curseur.update(screen)
        # update the music (to keep it running if it reaches the end)
        Utils.update_music_menu()
        pygame.display.flip()
        # deal with the events
        for event in pygame.event.get():
            if event.type ==  QUIT:
                clickSound.play()
                pygame.mixer.music.fadeout(200)
                clock.tick(5)
                running = 0
                return False
            elif event.type == MOUSEBUTTONDOWN and event.button == 1 :
                if back.highlighten :
                    clickSound.play()
                    running = 0
                    return True
                else :
                    for i, button in enumerate(buttons):
                        if button.highlighten:
                            if i >= n_customs : # si le niveau est un niveau custom :
                                clickSound.play()
                                pygame.mixer.music.fadeout(200)
                                clock.tick(5)
                                main(i - n_customs + 1, custom=True, start=True, from_selection=True)
                                if not pygame.mixer.get_init():
                                    return False
                                pygame.display.set_caption("Tank Game")
                            else:
                                if unlocked[i]:
                                    clickSound.play()
                                    pygame.mixer.music.fadeout(200)
                                    clock.tick(5)
                                    if main(i + 1, custom=False, start=True, from_selection=True):
                                        unlocked = get_unlocked()
                                    if not pygame.mixer.get_init():
                                        return False
                                    pygame.display.set_caption("Tank Game")

if __name__ == '__main__':
    level_selection_menu()
